frontend/src/metabase/admin/settings/components/widgets/AdminSettingInput.tsx

```diff
@@ -0,0 +1,206 @@
+import { jt, t } from "ttag";
+
+import { useAdminSetting } from "metabase/api";
+import { useDocsUrl, useToast } from "metabase/common/hooks";
+import ExternalLink from "metabase/core/components/ExternalLink";
+import type { GenericErrorResponse } from "metabase/lib/errors";
+import {
+  Box,
+  type BoxProps,
+  Radio,
+  Select,
+  Switch,
+  TextInput,
+  Textarea,
+} from "metabase/ui";
+import type { SettingKey } from "metabase-types/api";
+
+import { SettingHeader } from "../SettingHeader";
+
+type OptionsInputType = "select" | "radio";
+type TextualInputType = "text" | "number" | "password" | "textarea";
+type BooleanInputType = "boolean";
+
+type InputDetails =
+  | {
+      inputType: TextualInputType;
+      options?: never;
+      placeholder?: string;
+    }
+  | {
+      inputType: OptionsInputType;
+      options: { label: string; value: string }[];
+      placeholder?: string;
+    }
+  | {
+      inputType: BooleanInputType;
+      options?: never;
+      placeholder?: never;
+    };
+
+export type AdminSettingInputProps<S extends SettingKey> = {
+  name: S;
+  title?: string;
+  description?: React.ReactNode;
+  hidden?: boolean;
+} & InputDetails &
+  BoxProps;
+
+/**
+ * A simple admin settings component for basic needs, if you need something special,
+ * create a special component (in the widgets/ folder) instead of building one-off
+ * features into this component
+ */
+export function AdminSettingInput<SettingName extends SettingKey>({
+  title,
+  description,
+  name,
+  inputType,
+  hidden,
+  placeholder,
+  options,
+  ...boxProps
+}: AdminSettingInputProps<SettingName>) {
+  const [sendToast] = useToast();
+
+  const {
+    value: initialValue,
+    updateSetting,
+    isLoading,
+    settingDetails,
+  } = useAdminSetting(name);
+
+  const handleChange = (newValue: string | boolean | number) => {
+    if (newValue === initialValue) {
+      return;
+    }
+    updateSetting({ key: name, value: newValue }).then(response => {
+      if (response?.error) {
+        const message =
+          (response.error as GenericErrorResponse)?.message ||
+          t`Error saving ${title}`;
+        sendToast({ message, icon: "check", toastColor: "danger" });
+      } else {
+        sendToast({ message: t`${title} changes saved`, icon: "check" });
+      }
+    });
+  };
+
+  if (hidden || isLoading) {
+    return null;
+  }
+
+  return (
+    <Box {...boxProps}>
+      <SettingHeader id={name} title={title} description={description} />
+      {settingDetails?.is_env_setting && settingDetails?.env_name ? (
+        <SetByEnvVar varName={settingDetails.env_name} />
+      ) : (
+        <AdminSettingInputComponent
+          name={name}
+          value={initialValue}
+          onChange={handleChange}
+          options={options}
+          placeholder={placeholder}
+          inputType={inputType}
+        />
+      )}
+    </Box>
+  );
+}
+
+export function AdminSettingInputComponent({
```

Comment:
Not sure about the `Component` suffix, seeing as everything is a Component. 

Also, does this need to be exported?

Reply:
yes it's helpful to export it, you can see how it's used in #55064. how about `AdminSettingInputPure` ? any other name suggestion?

Reply:
AdminSettingAdminSettingInputComponentBasic